Functions/Gherkin.Core.ps1

```diff
@@ -0,0 +1,1304 @@
+
+        .PARAMETER NoMultiline
+            Controls whether or not Step Definition DocString and DataTable arguments are printed to the
+            console during the test run.
+
+        .PARAMETER Expand
+            Controls whether or not Scenario Outline example scenarios are displayed fully expanded instead of
+            just printing the scenario outline and table. This can be useful to determine exactly which step
+            in a scenario outline example scenario failed.
+
+        .PARAMETER PassThru
+            Returns a custom object (PSCustomObject) that contains the test results.
+            By default, Invoke-Gherkin writes to the host program, not to the output stream (stdout).
+            If you try to save the result in a variable, the variable is empty unless you
+            use the PassThru parameter.
+            To suppress the host output, use the Quiet parameter.
+
+        .EXAMPLE
+            Invoke-Gherkin
+
+            This will find all *.feature specifications and execute their tests. No exit code will be returned and no log file will be saved.
+
+        .EXAMPLE
+            Invoke-Gherkin -Path ./tests/Utils*
+
+            This will run all *.feature specifications under ./Tests that begin with Utils.
+
+        .EXAMPLE
+            Invoke-Gherkin -ScenarioName "Add Numbers"
+
+            This will only run the Scenario named "Add Numbers"
+
+        .EXAMPLE
+            Invoke-Gherkin -EnableExit -OutputXml "./artifacts/TestResults.xml"
+
+            This runs all tests from the current directory downwards and writes the results according to the NUnit schema to artifacts/TestResults.xml just below the current directory. The test run will return an exit code equal to the number of test failures.
+
+        .EXAMPLE
+            Invoke-Gherkin -CodeCoverage 'ScriptUnderTest.ps1'
+
+            Runs all *.feature specifications in the current directory, and generates a coverage report for all commands in the "ScriptUnderTest.ps1" file.
+
+        .EXAMPLE
+            Invoke-Gherkin -CodeCoverage @{ Path = 'ScriptUnderTest.ps1'; Function = 'FunctionUnderTest' }
+
+            Runs all *.feature specifications in the current directory, and generates a coverage report for all commands in the "FunctionUnderTest" function in the "ScriptUnderTest.ps1" file.
+
+        .EXAMPLE
+            Invoke-Gherkin -CodeCoverage @{ Path = 'ScriptUnderTest.ps1'; StartLine = 10; EndLine = 20 }
+
+            Runs all *.feature specifications in the current directory, and generates a coverage report for all commands on lines 10 through 20 in the "ScriptUnderTest.ps1" file.
+
+        .LINK
+            Invoke-Pester
+            https://kevinmarquette.github.io/2017-03-17-Powershell-Gherkin-specification-validation/
+            https://kevinmarquette.github.io/2017-04-30-Powershell-Gherkin-advanced-features/
+    #>
+    [CmdletBinding(DefaultParameterSetName = 'Default')]
```

Comment:
Where is the *Default* parameterset name defined? None of the parameters below specify it in the `Parameter` attribute definition.

Were you referring instead to **__AllParameterSets**?

Reply:
This code is making use of some standard behavior of PowerShell. Because the parameter `FailedLast` is explicitly added to the `RetestFailed` parameter set, that's the only parameter set the parameter is added to. So far, so good, this is standard PowerShell. Now, also standard with PowerShell, any parameter without a specific parameter set assignment is added to _all_ parameter sets defined on the Cmdlet. So, now the `RetestFailed` parameter set includes _all_ parameters for `Invoke-Gherkin`. Now in the `CmdletBinding` attribute, `Invoke-Gherkin` declares that the default parameter set is named `Default`. Even though no parameters are explicitly assigned to that parameter set, as previously mentioned, all parameters not explicitly assigned to a parameter set are assigned to every parameter set defined by the Cmdlet. The very fact that we said the default parameter set name is `Default` implies that there is a `Default` parameter set and since all other parameters other than `FailedLast` have no explicit parameter set mapping, these parameters are added to both the `Default` and `Retestfailed` parameter sets. I actually didn't think that doing this would work (and never really bothered to examine this before since `Invoke-Gherkin` just works).  This is a fairly neat short-hand trick to creating a parameter set.

Anyway, check it out for yourself, load Pester in a PowerShell ISE session and check out the help in the side-pane. Pretty neat stuff.

Reply:
BTW, please also note that I did not add this default parameter set name. Again, the code in the file moved, making it look as if this code was added by me.

Reply:
@fourpastmidnight Sounds good to me :)